packages/server/src/fhir/search.ts

```diff
@@ -1353,94 +1361,134 @@ function usesReferenceLookupTable(repo: Repository): boolean {
+  let innerQuery: SelectQuery;
+  if (link.details.type === SearchParameterType.CANONICAL) {
+    innerQuery = new SelectQuery(currentTable).whereExpr(
+      getCanonicalJoinCondition(selectQuery.tableName, link, currentTable)
+    );
+  } else {
+    innerQuery = new SelectQuery(currentTable).whereExpr(
+      lookupTableJoinCondition(selectQuery.tableName, link, currentTable)
+    );
+    currentTable = linkLiteralReference(innerQuery, currentTable, link);
+  }
+
+  // Add joins to inner query for all subsequent chain links
+  for (let i = 1; i < param.chain.length; i++) {
+    link = param.chain[i];
     if (link.details.type === SearchParameterType.CANONICAL) {
-      currentTable = linkCanonicalReference(selectQuery, currentTable, link);
+      currentTable = linkCanonicalReference(innerQuery, currentTable, link);
     } else {
-      currentTable = linkLiteralReference(selectQuery, currentTable, link, currentResourceType);
-    }
-    currentResourceType = link.resourceType;
-
-    if (link.filter) {
-      return buildSearchFilterExpression(
-        repo,
-        selectQuery,
-        currentResourceType as ResourceType,
-        currentTable,
-        link.filter
-      );
+      const lookupTable = linkReferenceLookupTable(innerQuery, currentTable, link);
+      currentTable = linkLiteralReference(innerQuery, lookupTable, link);
     }
   }
-  throw new OperationOutcomeError(badRequest('Unterminated chained search'));
+
+  // Add terminal conditions on final target table, and return EXISTS() over subquery
+  innerQuery
+    .where(new Column(currentTable, 'id'), '!=', null)
+    .whereExpr(
+      buildSearchFilterExpression(repo, innerQuery, link.targetType as ResourceType, currentTable, param.filter)
+    );
+  return new SqlFunction('EXISTS', [innerQuery]);
 }
 
+/**
+ * Join a query to the next table via canonical reference (i.e. by `url`).
+ * @param selectQuery - The query to which the join will be added.
+ * @param currentTable - The "current" table in the chained search construction.
+ * @param link - The current link of the chained search.
+ * @returns The next table alias.
+ */
 function linkCanonicalReference(selectQuery: SelectQuery, currentTable: string, link: ChainedSearchLink): string {
-  const nextTableAlias = selectQuery.getNextJoinAlias();
-  const eq = link.details.array ? 'IN_SUBQUERY' : '=';
-
-  let join: Condition;
-  if (link.reverse) {
-    join = new Condition(new Column(currentTable, 'url'), eq, new Column(nextTableAlias, link.details.columnName));
-  } else {
-    join = new Condition(new Column(nextTableAlias, 'url'), eq, new Column(currentTable, link.details.columnName));
-  }
-
-  selectQuery.join('LEFT JOIN', link.resourceType, nextTableAlias, join);
-  return nextTableAlias;
+  const nextTable = selectQuery.getNextJoinAlias();
+  const join = getCanonicalJoinCondition(currentTable, link, nextTable);
+  selectQuery.join('LEFT JOIN', nextChainedTable(link), nextTable, join);
```

Comment:
That doesn't work with `_filter` queries using `OR`: chained search requires `LEFT JOIN` to support that use case

Reply:
I think the LEFT JOIN is ensuring that all records from the main table remain in the output dataset regardless of whether any subsequent table can match content criteria.

An inner join definitely would change the semantics of the query. It would pre-filter some rows before you get to the WHERE clause.

I also think we can keep the original query simple and fast by simply dropping the distinct call without doing any further modifications (See my update on #5499).
I think that the Table.id for this kind of query has to be unique already as long as we keep the LEFT JOINs in place and the id column is constrained to uniqueness. 

Let me know if I am misunderstanding.